[Sema] Implement type checking for the `_typeEraser` attribute

hborla opened the pull request:
The rules for `@_typeEraser` are:
1. The type eraser must be a concrete nominal type
2. The type eraser must not have more restrictive access than the annotated protocol
3. The type eraser must conform to the annotated protocol
4. The type eraser must have an initializer of the form `init<T: AnnotatedProtocol>(erasing: T)`
   - Other generic requirements are okay as long as the init can always be called with a value of any type conforming to the protocol
   - The init cannot have more restrictive access than the annotated protocol

hborla committed:
[Sema] Implement type checking for the typeEraser attribute.

hborla committed:
[TypeCheckAttr] Allow a type eraser initializer to have more than one
generic requirement.

hborla committed:
[TypeCheckAttr] Provide better error messages for a typeEraser that
has invalid init candidates.

hborla committed:
[TypeCheckAttr] Implement access control checking for the typeEraser
attribute.

hborla committed:
[TypeCheckAttr] Improve diagnostics and candidate notes for typeEraser in
the case where the type eraser has no viable initializers.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2376,6 +2376,176 @@ void AttributeChecker::visitDynamicReplacementAttr(DynamicReplacementAttr *attr)
+  // The type eraser must conform to the annotated protocol
+  SmallVector<ProtocolConformance *, 2> conformances;
+  if (!nominalTypeDecl->lookupConformance(dc->getParentModule(), protocol,
+                                          conformances)) {
```
I'm not actually sure whether that's the best way to check conformance. I remember @DougGregor or somebody else told me that a better option would be `TypeChecker::conformsToProtocol`. /cc @slavapestov

slavapestov replied:
Yeah, it's a bit confusing because there are multiple layers. The API you're using here does not check conditional requirements or record incremental dependencies. I think inside the type checker, `TypeChecker::conformsToProtocol()` is your best bet, because it will do both of those things (except I'm not sure conditional requirements make sense in this case).

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2376,6 +2376,176 @@ void AttributeChecker::visitDynamicReplacementAttr(DynamicReplacementAttr *attr)
+    ParamDecl *param = *init->getParameters()->begin();
+    if (param->getArgumentName() != ctx.Id_erasing ||
+        !param->getInterfaceType()->isEqual(genericParamType) ||
+        !genericSignature->conformsToProtocol(genericParamType, protocol))
```
I think here as well, it might be better to use `TypeChecker::conformsToProtocol` call...

slavapestov replied:
No, this one is correct. GenericSignature::conformsToProtocol() tells you if a generic parameter in a signature has the given protocol requirement.

If you just call conformsToProtocol() with a GenericTypeParamType it won't have that information -- there's no associated signature.

xedin replied:
Thank you!

xedin reviewed (approved):
Looks great!

theblixguy commented on test/attr/typeEraser.swift:
```diff
@@ -1,19 +1,111 @@
+  init?<T: E2>(erasing t: T) {} // expected-note {{'init(erasing:)' cannot be failable}}
+}
+@_typeEraser(FailableInit) // expected-error {{type eraser 'FailableInit' has no viable initializer of the form 'init<T: E2>(erasing: T)'}}
+protocol E2 {}
```
Two tests that might be worth adding:

1. The type passed to `@_typeEraser` is a typealias.
2. The generic parameter `T` is on the parent type but used in the initializer.

hborla replied:
Done!

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2376,6 +2376,176 @@ void AttributeChecker::visitDynamicReplacementAttr(DynamicReplacementAttr *attr)
+        typeEraser->getContextSubstitutionMap(nominalTypeDecl->getParentModule(),
+                                              nominalTypeDecl);
+    QuerySubstitutionMap getSubstitution{baseMap};
+    auto subMap = genericSignature->getIdentitySubstitutionMap().subst(
```
I think this is just SubstitutionMap::get(...) with the same callback arguments

hborla replied:
Thank you!

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -2054,6 +2054,25 @@ class DifferentiableAttributeTypeCheckRequest
+  friend SimpleRequest;
+
+  // Evaluation
+  llvm::Expected<bool> evaluate(Evaluator &evaluator, TypeEraserAttr *attr,
```
Depending on what you're planning on doing with this request, it might be better if the result was the actual viable ConstructorDecl if one was found, instead of a boolean. Otherwise aren't you going to have to look up to find the candidate again?

hborla replied:
This is how I implemented it at first (even in the first commit of this PR), but when I extended the implementation to allow arbitrary generic requirements (as long as they're satisfied), I didn't want to re-implement overload resolution logic. I want the constraint system to handle that part

slavapestov replied:
Do you mean you're planning on building an Expr and passing it to typeCheckExpression() rather than building a fully-type checked AST for the call? I'm worried that with this approach the constraint system might pick a different overload than the one that this request picks. Also in the past we've had issues where checking this sub-constraint system could fail, which would create weird error cases that were hard to recover from. Today CSApply and ClangImporter mostly build type-checked AST for this reason.

It sounds like you already have everything you need to form the call, because you're building the SubstitutionMap for checking the generic requirements. Other than that, you would build a DeclRefExpr wrapped in a CallExpr.

But it's totally up to you, I won't press the issue further :)

slavapestov replied:
Oh, I get it, you have multiple viable candidates, and you want the constraint solver to pick one using the existing ranking rules. The attribute checker just has to guarantee that at least one viable candidate exists, not that they're unique.

Ok, that's a tricky one. Never mind.

hborla committed:
[TypeCheckAttr] Use the correct APIs for checking protocol conformance
and replacing type parameters in a generic signature when type checking
the typeEraser attribute.

hborla committed:
[TypeCheckAttr] Allow a typealias to be a type eraser and add additional
typeEraser tests.

hborla commented:
@swift-ci please smoke test
